content/en/blog/_posts/2023-09-15-Bootstrap-an-Air-Gapped-Cluster-with-Kubeadm/index.md

```diff
@@ -0,0 +1,665 @@
+
+**Start and enable Docker:**
+
+```bash
+systemctl restart docker
+systemctl enable docker
+```
+
+**Start and enable the kubelet:**
+
+```bash
+systemctl start kubelet
+systemctl enable kubelet
+```
+
+The container image registry that runs in Docker is only required for any CNI related containers and subsequent workload containers. This registry is **not** used to house the Kubernetes component containers. Note, nerdctl would have also worked here as an alternative to Docker and would have allowed for direct interaction with containerd. Docker was chosen for its familiarity.
+
+**Start Docker registry**
+
+```bash
+docker load -i registry_2.8.2.tar
+docker run -d -p 5000:5000 --restart=always --name registry registry:2.8.2
+```
+
+**Load Flannel containers into the Docker registry:**
+
+```bash
+docker load -i flannel_flannel_v0.22.0.tar
+docker load -i flannel_flannel-cni-plugin_v1.1.2.tar
+docker tag flannel/flannel:v0.22.0 localhost:5000/flannel/flannel:v0.22.0
+docker tag flannel/flannel-cni-plugin:v1.1.1 localhost:5000/flannel/flannel-cni-plugin:v1.1.1
+docker push localhost:5000/flannel/flannel:v0.22.0
+docker push localhost:5000/flannel/flannel-cni-plugin:v1.1.1
+```
+
+Load container images for Kubernetes components, via ctr:
+
+```bash
+images_files=(
+    "registry.k8s.io/kube-apiserver:${KUBE_RELEASE}"
+    "registry.k8s.io/kube-controller-manager:${KUBE_RELEASE}"
+    "registry.k8s.io/kube-scheduler:${KUBE_RELEASE}"
+    "registry.k8s.io/kube-proxy:${KUBE_RELEASE}"
+    "registry.k8s.io/pause:3.9"
+    "registry.k8s.io/etcd:3.5.7-0"
+    "registry.k8s.io/coredns/coredns:v1.10.1"
+    
+)
+
+
+for index in "${!image_files[@]}"; do
+
+    if [[-f "${image_files[$index]}" ]]; then
+
+        # The below line loads the images where they need to be on the VM
+        ctr -n k8s.io images import ${image_files[$index]}
+
+    else
+
+        echo "File ${image_files[$index]} not found!"
+
+    fi
+
+done
+```
+
+A totally reasonable question here could be "Why not use the Docker registry that was just stood up to house the K8s component images?" This simply didn't work even with the proper modification to the configuration file that gets passed to kubeadm.
+
+**Spin up the Kubernetes cluster**
+
+Check if a cluster is already running and tear it down if it is:
+
+```bash
+if systemctl is-active --quiet kubelet; then
+
+    # Reset the Kubernetes cluster
+
+    echo "A Kubernetes cluster is already running. Resetting the cluster..."
+
+    kubeadm reset -f
+
+fi
+```
+
+**Log into the Docker registry from inside the air-gapped VM:**
+
+```bash
+# OK for a demo; use secure credentials in production!
+
+DOCKER_USER=user
+DOCKER_PASS=pass
+echo ${DOCKER_PASS} | docker login --username=${DOCKER_USER} --password-stdin localhost:5000
+```
+
+**Create a cluster configuration file and initialize the cluster:**
+
+```bash
+echo "---
+
+apiVersion: kubeadm.k8s.io/v1beta3
```

Comment:
this configuration includes defaults that can confuse the user. better omit them.

clusterName, certificateDir, etcd, ...

Reply:
Seems OK to leave in `clusterName` - it's a hint to readers that they might want to set something.

Reply:
I'm happy to remove anything that is a default value.